2-WebApp-graph-user/2-5-HybridFlow/Program.cs

```diff
@@ -0,0 +1,73 @@
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc.Authorization;
+using Microsoft.Identity.Web;
+using WebApp_OpenIDConnect_DotNet.Services;
+
+var serviceCollection = new ServiceCollection();
+serviceCollection.AddSingleton<IConfidentialClientApplicationService, ConfidentialClientApplicationService>();
+
+var serviceProvider = serviceCollection.BuildServiceProvider();
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.Add(serviceCollection.Single());
+
+builder.Services.Concat(serviceCollection);
+
+builder.Services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
+    .AddMicrosoftIdentityWebApp(options =>
+    {
+        options.ResponseType = "code";
+        options.Events.OnAuthorizationCodeReceived = async context =>
+        {
+            context.TokenEndpointRequest.Parameters.TryGetValue("code_verifier", out var codeVerifier);
+            context.HandleCodeRedemption();
+
+            var clientService = serviceProvider.GetService<IConfidentialClientApplicationService>();
+
+            var authResult = await clientService.GetAuthenticationResultAsync(context.ProtocolMessage.Code, codeVerifier);
+            var spaCode = authResult.SpaAuthCode;
+            context.Response.Cookies.Append("Microsoft.Identity.Hybrid.Authentication.Cookie", spaCode);
```

Comment:
Instead of a cookie, I would recommend we render this to JavaScript using a template.

Reply:
Disregard the above.

I think I found something. I'll put it up in a bit here.

Reply:
Coming back to this,

I really wanted to store the code in a session that could be accessed server side and add some more obfuscation but was having some issues setting that up with the `OnAuthorizationCodeReceived` method within the `AddMicrosoftIdentityWebApp` of the main application.

Instead, I took your suggestion and had the server pre-render JavaScript with an IIFC in `_Layout.cshtml` that exchanges the code for an auth token by reading the cookie. However, the server takes on the responsibility of erasing the cookie and configuring the IIFC based on whether or not the cookie is set.

Reply:
@jasonnutter 

Never I mind. I managed to stuff the value in a `Session` object where to be encrypted and decrypted server side with the ASP.NET session. The cookie is gone.